featuremanagement/_featuremanager.py

```diff
@@ -283,31 +306,32 @@ def _check_feature_filters(self, feature_flag, targeting_context, **kwargs):
             elif self._filters[filter_name].evaluate(feature_filter, **kwargs):
                 evaluation_event.enabled = True
                 break
-        return evaluation_event
 
-    def _assign_allocation(self, feature_flag, evaluation_event, targeting_context):
-        if feature_flag.allocation and feature_flag.variants:
-            variant_name = self._assign_variant(feature_flag, targeting_context)
-            if variant_name:
-                evaluation_event.enabled = FeatureManager._check_variant_override(
-                    feature_flag.variants, variant_name, evaluation_event.enabled
-                ).enabled
-                evaluation_event.variant = self._variant_name_to_variant(feature_flag, variant_name)
-                evaluation_event.feature = feature_flag
-                return evaluation_event
-
-        variant_name = None
-        if evaluation_event.enabled:
-            evaluation_event = FeatureManager._check_default_enabled_variant(feature_flag)
-            if feature_flag.allocation:
-                variant_name = feature_flag.allocation.default_when_enabled
-        else:
-            evaluation_event = FeatureManager._check_default_disabled_variant(feature_flag)
-            if feature_flag.allocation:
-                variant_name = feature_flag.allocation.default_when_disabled
-        evaluation_event.variant = self._variant_name_to_variant(feature_flag, variant_name)
-        evaluation_event.feature = feature_flag
-        return evaluation_event
+    def _assign_allocation(self, evaluation_event, targeting_context):
+        feature_flag = evaluation_event.feature
+        if feature_flag.variants:
+            if not feature_flag.allocation:
+                if evaluation_event.enabled:
+                    evaluation_event.reason = VariantAssignmentReason.DEFAULT_WHEN_ENABLED
+                    return
+                evaluation_event.reason = VariantAssignmentReason.DEFAULT_WHEN_DISABLED
+                return
+            if not evaluation_event.enabled:
+                FeatureManager._check_default_disabled_variant(evaluation_event)
+                evaluation_event.variant = self._variant_name_to_variant(
```

Comment:
We can add a shortcut in `_variant_name_to_variant` when `variant_name` is None. For example, if `feature_flag.allocation.default_when_enabled` is `None` here, the for loop in Line 206 can be skipped.